proposals/user-scripts-execute-api.md

```diff
@@ -0,0 +1,148 @@
+
+### User Scripts API
+
+The User Scripts API was introduced in MV3 to register user scripts with arbitrary code, specifically targeting user script managers' use cases. User scripts are injected based on url matching through match patterns and globs. The API requires a new permission, and allows each browser to implement their own restrictions.
+In the User Script API proposal, executing a user script one time was [presented](https://github.com/w3c/webextensions/blob/main/proposals/user-scripts-api.md#execute-user-scripts-one-time) as a potential future enhancement.
+
+## API Overview
+
+### Objective
+
+Allow extensions to inject user scripts programmatically into web content given a target context.
+
+#### Use Cases
+
+The user wants to inject a user script on the page currently loaded in a given tab one time, but does not want to inject the script on that site automatically in the future. That is, the parameter that determines the injection is a tab ID and not based on origin.
+
+### Availability: Public
+
+This API will be public to all extensions (with the appropriate permission).
+
+### Consumers
+
+Any existing user script manager consumers of `chrome.tabs.executeScript()` that inject user scripts programmatically are intended to migrate to this new version in Manifest V3.
+
+## Implementation
+
+### API Schema
+
+#### Types
+
+```
+dictionary UserScriptInjection {
+  // Whether the injection should be triggered in the target as soon as
+  // possible. Note that this is not a guarantee that injection will occur
+  // prior to page load, as the page may have already loaded by the time the
+  // script reaches the target.
+  injectImmediately?: boolean,
+  // The script source to inject into the target.
+  js: ScriptSource,
+  // Details specifying the target into which to inject the script.
+  target: InjectionTarget,
+  // The JavaScript "world" to run the script in. The default is `USER_SCRIPT`.
+  world?: ExecutionWorld
+}
+
+dictionary InjectionTarget {
+  // Whether the script should inject into all frames within the tab. Defaults
+  // to false. This must not be true if `frameIds` is specified.
+  allFrames?: boolean,
+  // The IDs of specific documentIds to inject into. This must not be set if
+  // frameIds is set.
+  documentIds?: string[]
+  // The IDs of specific frames to inject into.
+  frameIds?: number[]
+  // The ID of the tab into which to inject.
+  tabId: number
+}
+
+dictionary InjectionResult {
+  // The document associated with the injection.
+  documentId: string,
+  // The frame associated with the injection.
+  frameId: number,
+  // The result of the script execution.
+  result?: any
```

Comment:
One `result` implies that there is an unambiguous way to encode the result. That is not the case if there are errors, ranging from runtime errors (while executing the script) to permission errors.

I reported a similar issue about `scripting.executeScript` before at https://issues.chromium.org/issues/40205757
Proposed fix: add an `error` property (MDN documentation with `error` property of scripting.executeScript: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/scripting/executeScript#return_value):

```suggestion
  // The result of the script execution.
  result?: any
  // The error, if any. error and result are mutually exclusive.
  error?: any
```

Reply:
Agreed; this has been something I've been thinking about in the executeScript() API for [awhile](https://source.chromium.org/chromium/chromium/src/+/main:chrome/test/data/extensions/api_test/scripting/main_frame/worker.js;l=275-277;drc=bf596c951f436be7f275553a09921064c2de9e93).

Reply:
This makes sense. Added to the proposal. We can also look to add this to `scripting.executeScript()`

Reply:
Taking a second look, should error be a string? In chrome, the error associated with the injection is a string (see [FrameResult](https://source.chromium.org/chromium/chromium/src/+/1b571bef02256072be9e289270d6067da1e0de73:extensions/browser/script_executor.h;l=88-90)). Would that work for other browsers?

Reply:
Oops, I missed this question, so answering now.

> Taking a second look, should error be a string? In chrome, the error associated with the injection is a string (see [FrameResult](https://source.chromium.org/chromium/chromium/src/+/1b571bef02256072be9e289270d6067da1e0de73:extensions/browser/script_executor.h;l=88-90)). Would that work for other browsers?

A string would be insufficient to catch user-defined "errors", whether from the last expression being `Promise.reject(12345)` or by an explicit `throw 12345`. For this reason, we have declared the type as `any`, and documented at https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/scripting/executeScript#error as follows:

> If an error occurs, contains the value the script threw or rejected with. Typically this is an error object with a message property but it could be any value (including primitives and undefined).